core/templates/pages/contributor-dashboard-page/modal-templates/translation-modal.controller.ts

```diff
@@ -100,6 +158,75 @@ angular.module('oppia').controller('TranslationModalController', [
       $scope.activeWrittenTranslation.html = '';
     };
 
+    $scope.findImgAttributes = function(elements, type) {
+      const textWrapperLength = 6;
+      let attributes = Array.from(elements, function(element: HTMLElement) {
+        if (element.localName === 'oppia-noninteractive-image') {
+          const attribute = element.attributes[type].value;
+          return attribute.substring(
+            textWrapperLength, attribute.length - textWrapperLength);
+        }
+      });
+      return attributes.filter(attribute => attribute);
+    };
+
+    $scope.findAvailableElements = function(
+        originalElements, translatedElements, isRawImage) {
+      let states = isRawImage ? Array.from(
+        originalElements, function(originalElement) {
+          return translatedElements.some(
+            translatedElement => (translatedElement === originalElement));
+        }) : Array.from(
+          originalElements, function(originalElement) {
+            return translatedElements.some(
+              translatedElement => (
+                translatedElement === originalElement &&
+                originalElement !== ''));
+          });
+      return states;
+    };
+
+    $scope.getTexts = function(rawText) {
+      const foundImageFilePaths = $scope.findImgAttributes(
+        rawText, 'filepath-with-value');
+      const foundImageAltTxts = $scope.findImgAttributes(
+        rawText, 'alt-with-value');
+      const foundImageDescriptions = $scope.findImgAttributes(
+        rawText, 'caption-with-value');
+
+      return {
+        foundImageFilePaths,
+        foundImageAltTxts,
+        foundImageDescriptions
+      };
+    };
+
+    $scope.validateImages = function(
+        textToTranslate, translatedText): TranslationError {
+      const translatedElements = $scope.getTexts(translatedText);
+      const originalElements = $scope.getTexts(textToTranslate);
+
+      const copiedImgStates = $scope.findAvailableElements(
+        originalElements.foundImageFilePaths,
+        translatedElements.foundImageFilePaths, true);
+      const duplicateImgAltTextStates = $scope.findAvailableElements(
+        originalElements.foundImageAltTxts,
+        translatedElements.foundImageAltTxts, false);
+      const duplicateImgDescriptionStates = $scope.findAvailableElements(
+        originalElements.foundImageDescriptions,
+        translatedElements.foundImageDescriptions, false);
+
+      const hasUncopiedImgs = copiedImgStates.some(state => state === false);
+      const hasDuplicateAltTexts = duplicateImgAltTextStates.some(
+        state => state === true);
+      const hasDuplicateDescriptions = duplicateImgDescriptionStates.some(
+        state => state === true);
```

Comment:
Instead of passing in a boolean to `findAvailableElements`, let's just have two separate functions. We can also use [`Array.every()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/every) instead of `some()` to combine this logic in the same function. E.g. the first function can look like:

```
$scope.didNotCopyAllElements = function(originalElements, translatedElements) {
  const hasMatchingTranslatedElement = (element) => translatedElements.includes(element);
  return originalElements.every(hasMatchingTranslatedElement);
};
``` 

Btw, I believe you don't need to call `Array.from()` again here as the supplied arguments will be arrays already.

Reply:
Done